### PR TITLE
pythonPackages.adafruit-nrfutil: init at 0.5.3.post17

### PR DESCRIPTION
**Pull request by StarGate01**
###### Description of changes

This PR adds the Python 3 adafruit-nrfutil package from PyPi. See https://github.com/adafruit/Adafruit_nRF52_nrfutil .

Depends on: https://github.com/adafruit/Adafruit_nRF52_nrfutil/pull/38 ~~(and a successive new version)~~. Required for the tests to pass under recent Python and `ecdsa` versions. Before that PR is merged and a new version is tagged, this PR will be a draft.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+, pyserial
+, click
+, ecdsa
+, behave
+}:
+
+buildPythonPackage rec {
+  pname = "adafruit-nrfutil";
+  version = "0.5.3.post16";
+
```
Comment (fabaff) [inline pkgs/development/python-modules/adafruit-nrfutil/default.nix:13]: I suggest to disable the module on older Python releases which are EOL as the chance is very low that somebody will test the module.

```suggestion
  version = "0.5.3.post16";
  format = "setuptools";

  disabled = pythonOlder "3.7";
```

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  propagatedBuildInputs = [
+    pyserial
+    click
+    ecdsa
+    behave
+  ];
+
+  # package does not include tests
+  doCheck = false;
+
```
Comment (fabaff) [inline pkgs/development/python-modules/adafruit-nrfutil/default.nix:28]: An additional safe-guard.

```suggestion

  pythonImportsCheck = [
    "nordicsemi"
  ];

```

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  meta = with lib; {
+    homepage = "https://github.com/adafruit/Adafruit_nRF52_nrfutil";
+    description = "Modified version of Nordic's nrfutil 0.5.x for use with the Adafruit Feather nRF52";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ stargate01 ];
+  };
+
+}
```
Comment (fabaff) [inline pkgs/development/python-modules/adafruit-nrfutil/default.nix:36]:
```suggestion
  };
}
```

File: maintainers/maintainer-list.nix [hunk 4]
```diff
@@ -14663,4 +14663,10 @@
     github = "npatsakula";
     githubId = 23001619;
   };
+  stargate01 = {
```
Comment (fabaff) [inline maintainers/maintainer-list.nix:14666]: This should be a separate commit in this PR.

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix (hunk 2 above)
Comment (StarGate01) [reply pkgs/development/python-modules/adafruit-nrfutil/default.nix:28]: I am quite new to Nix packaging - how does this change help? The unit tests should eventually cover the testing (?)

**Comment by StarGate01** [comment]
This PR now depends on https://github.com/adafruit/Adafruit_nRF52_nrfutil/pull/38 . The SHA in the package definition is intentially invalid, as the next version (`post18`) does not exist yet.

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix (hunk 2 above)
Comment (fabaff) [reply pkgs/development/python-modules/adafruit-nrfutil/default.nix:42]: > The unit tests should eventually cover the testing (?)

Unit tests usually import the module but if the tests are not executed then `pythonImportsCheck` helps to ensure that the module can be imported.

It is a cheap check. It happens quite often that the tests have to be disabled due to whatever reason during the life-time of a package.

Comment (StarGate01) [reply pkgs/development/python-modules/adafruit-nrfutil/default.nix:42]: Thank you for the explanation !

**Comment by StarGate01** [comment]
@fabaff The [fixed tests have been merged](https://github.com/adafruit/Adafruit_nRF52_nrfutil/pull/38) into the upstream Adafruit repository. An upgrade of the version number was discussed, but ultimately deemed unnecessary.

The package now properly runs the included tests.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 178558` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.adafruit-nrfutil</li>
    <li>python39Packages.adafruit-nrfutil</li>
  </ul>
</details>

**Comment by StarGate01** [comment]
@fabaff What does `nixpkgs-review` tell me? I am kind of new to that tool. Should I ensure Python 37, 38, 39, 310 and 311 compatibility?

**Comment by SuperSandro2000** [comment]
> Should I ensure Python 37, 38, 39, 310 and 311 compatibility?

Right now we only really care about 39 and 310. If 311 is more widely adopted in the ecosystem then eventually that one, too.

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  pname = "adafruit-nrfutil";
+  version = "0.5.3.post17";
+  format = "setuptools";
+
+  disabled = pythonOlder "3.7";
+
+  src = fetchFromGitHub {
+    owner = "adafruit";
+    repo = "Adafruit_nRF52_nrfutil";
+    rev = "1361059009ff6a24d63b37eb3a4b28127837ead2";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/adafruit-nrfutil/default.nix:22]:
```suggestion
    rev = version;
```

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    click
+    ecdsa
+  ];
+
+  checkInputs = [
+    behave
+    nose
+  ];
+
+  postPatch = ''
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/adafruit-nrfutil/default.nix:37]:
```suggestion
  preCheck = ''
```

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    mkdir test-reports
+  '';
+
+  doCheck = true;
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/adafruit-nrfutil/default.nix:42]:
```suggestion
```
that is the default

File: pkgs/development/python-modules/adafruit-nrfutil/default.nix (hunk 5 above)
Comment (StarGate01) [reply pkgs/development/python-modules/adafruit-nrfutil/default.nix:22]: The tag `0.5.3.post17` does not point to a version containing the fixes for the required tests (https://github.com/adafruit/Adafruit_nRF52_nrfutil/pull/38/commits/e5fbcc8ee5958041db38c04139ba686bf7d1b845). The maintainer decided against releasing a new version, see https://github.com/adafruit/Adafruit_nRF52_nrfutil/pull/38#issuecomment-1169876952 .

Comment (SuperSandro2000) [reply pkgs/development/python-modules/adafruit-nrfutil/default.nix:22]: We can't just fetch a newer commit and leave the version as is. I would suggest to download the tagged version and fetch the patch we need.

Comment (StarGate01) [reply pkgs/development/python-modules/adafruit-nrfutil/default.nix]: Fair point. I adjusted the code to download the PR.

**Comment by SuperSandro2000** [comment]
@ofborg build python310Packages.adafruit-nrfutil
